smolyay/basis/basis.py

```diff
@@ -0,0 +1,249 @@
+import abc
+import math
+
+class BasisFunction(abc.ABC):
+    """Abstract class for basis functions.
+
+    Creates a framework for basis functions that are the building blocks
+    for the terms in the surrogate model. 
+    ``points`` property represents all the unique 1D points of the 
+    basis function associated with a Smolyak index described by the 
+    class IndexGrid.
+
+    """
+
+    def __init__(self):
+        self._points = []
+
+    @property
+    def points(self):
+        """The points of the basis function assigned to Smolyak indices"""
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @abc.abstractmethod
+    def _compute_points(self):
+        """Compute the value of the points for the basis function
+        Computes the 1D points associated with the basis function, The
+        method of computation depends on the basis function and what
+        the points are meant to represent
+        """
+        pass
+
+
+    @abc.abstractmethod
+    def __call__(self,x):
+        """Compute term of basis function
+        Returns the output of the basis function with input x
+
+        Parameters
+        ----------
+        x : float
+            input
+
+        Returns
+        -------
+        y : float
+            output
+        """
+        pass
+
+
+class ChebyshevFirstKind(BasisFunction):
+    r"""Basis function Chebyshev polynomials of the first kind
+    Chebyshev polynomials are a sequence of polynomials described by the
+    following recurrence relation:
+
+    ..math:
+        T_0(x) = 1
+        T_1(x) = x
+        T_{n+1}(x) = 2xT_n(x) - T_{n-1}(x)
+
+    ``n`` is the degree of the polynomial this object describes
+    ``points`` represents the extrema of the polynomial of this degree on
+    the domain [-1,1]
+
+    The extrema of the polynomial are calculated via the following
+    equation:
+
+    ..math:
+        x_{n,j}^* = -\cos\left((\frac{j-1}{n-1}\pi\right), j = 1,...,n
+        n = 2^{exactness-1}+1
+
+    Parameters
+    ----------
+    n : int
+        degree of the Chebyshev polynomial
+    """
+
+    def __init__(self,n):
+        super().__init__()
+        self._n = n
+
+    @property
+    def points(self):
+        """extrema of polynomial"""
+        if len(self._points) == 0:
+            self._compute_points()
+        return self._points
+
+    @property
+    def n(self):
+        """degree of polynomial"""
+        return self._n
+
+    @n.setter
+    def n(self,n):
+        if self._n != n:
+            self._n = n
+            self._points = []
+
+    def _compute_points(self):
+        """Compute extrema of Chebyshev polynomial of the first kind"""
+        if self._n == 0:
+            self._points = [0]
+        else:
+            for i in range(0,self._n+1):
+                temp = round(-math.cos(math.pi*(i)/(self._n)),15) + 0
```

Comment:
The rounding is to ensure that anything close to zero like 1e-16 is rounded down to zero and values like 0.9999999999 is seen as 1. The + 0 part causes -0 and +0 float values to be read as a plain 0. 
This ensures that the code correctly recognizes duplicate extrema values as duplicates.

Reply:
Ah, so that should be the job of whatever code detects duplicates, and not the basis function itself. I would recommend you turn this into the normal math (no rounding), then use `numpy.isclose` with some specified tolerance to detect duplicates when this is actually needed.